dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NormalizeLoops.java

```diff
@@ -69,5 +104,46 @@ public void run(Function function) {
 
             counter++;
         }
+
+        // Guarantees header is the only entry point
+        for (Label label : function.getEvents(Label.class)) {
+            final List<CondJump> backJumps = label.getJumpSet().stream()
+                    .filter(j -> j.getLocalId() > label.getLocalId())
+                    .sorted()
+                    .toList();
+            final Label loopBegin = label;
+            final CondJump uniqueBackJump = backJumps.get(0);
+
+            final List<Label> loopBodyLabels = loopBegin.getSuccessor().getSuccessors().stream()
+                    .takeWhile(ev -> ev != uniqueBackJump)
+                    .filter(Label.class::isInstance).map(Label.class::cast)
+                    .toList();
+            for (Label l : loopBodyLabels) {
+                final List<CondJump> externalEntries = l.getJumpSet().stream()
+                        .filter(j -> j.getLocalId() < loopBegin.getLocalId() ||
+                                j.getLocalId() > uniqueBackJump.getLocalId())
+                        .toList();
+
+                for (CondJump fromOutside : externalEntries) {
+
+                    final Register jumpedFromOutside = function.newRegister("__jumpedFromOutside",
+                            types.getBooleanType());
+                    final Local setJumpedFromOutside = EventFactory.newLocal(jumpedFromOutside, expressions.makeTrue());
+                    final CondJump jumpToHeader = EventFactory.newGoto(loopBegin);
+                    final Label internalLabel = fromOutside.getLabel();
+
+                    fromOutside.replaceBy(setJumpedFromOutside);
+                    setJumpedFromOutside.insertAfter(jumpToHeader);
```

Comment:
`replaceBy` does not copy over metadata. Use `.updateReferences(Map.of(<originalLabel>, <loopHeader>))` to retarget the external jump to the header while retaining it's original metadata.